<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
               "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
lang="en" xml:lang="en">
<head>
<title>Starter cookbook to processing genome w/ annotation pacakges for NCBI</title>
<meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1"/>
<meta name="generator" content="Org-mode"/>
<meta name="generated" content="2014-09-08 15:59:53 PDT"/>
<meta name="author" content="Katherine Amrine"/>
<meta name="description" content=""/>
<meta name="keywords" content=""/>
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  html { font-family: Times, serif; font-size: 12pt; }
  .title  { text-align: center; }
  .todo   { color: red; }
  .done   { color: green; }
  .tag    { background-color: #add8e6; font-weight:normal }
  .target { }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  p.verse { margin-left: 3% }
  pre {
	border: 1pt solid #AEBDCC;
	background-color: #F3F5F7;
	padding: 5pt;
	font-family: courier, monospace;
        font-size: 90%;
        overflow:auto;
  }
  table { border-collapse: collapse; }
  td, th { vertical-align: top; }
  dt { font-weight: bold; }
  div.figure { padding: 0.5em; }
  div.figure p { text-align: center; }
  .linenr { font-size:smaller }
  .code-highlighted {background-color:#ffff00;}
  .org-info-js_info-navigation { border-style:none; }
  #org-info-js_console-label { font-size:10px; font-weight:bold;
                               white-space:nowrap; }
  .org-info-js_search-highlight {background-color:#ffff00; color:#000000;
                                 font-weight:bold; }
  /*]]>*/-->
</style>
<script type="text/javascript">
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">

<h1 class="title">Starter cookbook to processing genome w/ annotation pacakges for NCBI</h1>

<p>Alrighty, here is my step-by-step and explanation for what I did to
do a WGS annotated genome submission for NCBI
</p>

<div id="outline-container-1" class="outline-2">
<h2 id="sec-1">Step 1 is to notify NCBI you will be submitting data </h2>
<div class="outline-text-2" id="text-1">



</div>

<div id="outline-container-1.1" class="outline-3">
<h3 id="sec-1.1">create a bioproject submission per genome. </h3>
<div class="outline-text-3" id="text-1.1">

<p>Specify that you will be submitting WGS, Annotated, SRA, etc. 
</p>
</div>

</div>

<div id="outline-container-1.2" class="outline-3">
<h3 id="sec-1.2">Next create a biosample submission </h3>
<div class="outline-text-3" id="text-1.2">

<p>include details on your sample
</p>
</div>

</div>

<div id="outline-container-1.3" class="outline-3">
<h3 id="sec-1.3">Next start the WGS submission process </h3>
<div class="outline-text-3" id="text-1.3">

<p>here, I think they will ask you for a prefix, and in parenthesis, it will say
"leave blank if you want NCBI to assign one for you". I am paraphrasing because
I didn't realize this was important until I had already missed this step. 
It is important because that prefix will be what you have to start all your gene
names with. When you get down to the <code>newIDs.new.ncbi.txt</code> file I ask for in Step3,
this prefix is what needs to go in this file. 
</p>
<p>
also, they will ask you for the type of evidence that you have for the strings of 
N's within sequences. First off, N's do NOT occur as singletons, so they are not 
ambiguous bases. Ns refer to gaps of the size of the string of Ns if they ask you. 
</p>
</div>

<div id="outline-container-1.3.1" class="outline-4">
<h4 id="sec-1.3.1">WHAT YOU WILL NEED </h4>
<div class="outline-text-4" id="text-1.3.1">

<p>-fasta file with scaffolds with concise names
this means that you really need to slim down names to as simple of
identifiers as possible. 
-gff version 2 file (with the scripts I provide here)
you don't need version 2, but I used it and wrote my scripts to
use this type of file. 
</p>
</div>

</div>

<div id="outline-container-1.3.2" class="outline-4">
<h4 id="sec-1.3.2">WHAT I GIVE YOU </h4>
<div class="outline-text-4" id="text-1.3.2">


</div>

<div id="outline-container-1.3.2.1" class="outline-5">
<h5 id="sec-1.3.2.1">a crappy script to do the first conversion of the gff file to the tbl </h5>
<div class="outline-text-5" id="text-1.3.2.1">

</div>

</div>

<div id="outline-container-1.3.2.2" class="outline-5">
<h5 id="sec-1.3.2.2">file required for the tbl2asn script that will ultimately make your  asn submission file </h5>
<div class="outline-text-5" id="text-1.3.2.2">

</div>

</div>

<div id="outline-container-1.3.2.3" class="outline-5">
<h5 id="sec-1.3.2.3">another crappy script to substitute in the names necessary </h5>
<div class="outline-text-5" id="text-1.3.2.3">


</div>
</div>
</div>
</div>

</div>

<div id="outline-container-2" class="outline-2">
<h2 id="sec-2">Step 2 is to pre-process the peptide files </h2>
<div class="outline-text-2" id="text-2">


</div>

<div id="outline-container-2.1" class="outline-3">
<h3 id="sec-2.1">isolate the peptide sequences with internal stops (check number of seqs) </h3>
<div class="outline-text-3" id="text-2.1">




<pre class="example">fasgrep -s "-" &lt;peptide_file.fas&gt; &gt; &lt;file_with_internal_stop_sequences.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-2.2" class="outline-3">
<h3 id="sec-2.2">remove peptide sequences with internal stops (confirm you are missing seqs) </h3>
<div class="outline-text-3" id="text-2.2">




<pre class="example">fasgrep -vs "-" &lt;peptide_file.fas&gt; &gt; &lt;peptiled_file_minus_internal_stops.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-2.3" class="outline-3">
<h3 id="sec-2.3">remove annotated genes with internal stops from gff file (confirm correct unique ids are removed) </h3>
<div class="outline-text-3" id="text-2.3">




<pre class="example">egrep -f &lt;(egrep "&gt;" &lt;file_with_internal_stop_sequences.fas&gt; | cut -c 2- |cut -f1,2 -d"." | perl -pe "s/\n/\\\&gt;\n/") &lt;original_gff_file.gff&gt; &gt; &lt;new_gff_minus_inernal_stop_peptides.gff&gt;
</pre>



</div>
</div>

</div>

<div id="outline-container-3" class="outline-2">
<h2 id="sec-3">Step 3 is to fix the scaffolds files </h2>
<div class="outline-text-2" id="text-3">


</div>

<div id="outline-container-3.1" class="outline-3">
<h3 id="sec-3.1">remove trailing ends (no one cares where these are, just truncate them) </h3>
<div class="outline-text-3" id="text-3.1">




<pre class="example">fassub -s 'N{1,}$' '' &lt;scaffolds_file_original.fas&gt; &gt; &lt;scaffolds_no_trailing.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-3.2" class="outline-3">
<h3 id="sec-3.2">find sequences with leading n's and their numbers, put into file </h3>
<div class="outline-text-3" id="text-3.2">




<pre class="example">fasgrep -s "^N" &lt;scaffolds_no_trailing.fas&gt; | perl -ne 'BEGIN{$start=0;}if(/^\&gt;/){$finish=0;chomp;$seqname=$_;$start=1;$length=0;}if($start==1 &amp;&amp; $finish==0){/(^N{1,})/;$length += length($1);if(/[ATCG]/){ print join"",$seqname,"\t",$length,"\n";$start=0;$finish=1;}}' | sort | uniq &gt; leading_Ns.txt
</pre>



</div>

</div>

<div id="outline-container-3.3" class="outline-3">
<h3 id="sec-3.3">change coordinates for leading N sequences in gff file </h3>
<div class="outline-text-3" id="text-3.3">

<p>for this script <i>changecoords.pl</i>, you need the file <code>leading_Ns.txt</code>
but it takes the gff file and fixes all the coordinates
</p>


<pre class="example">./change_coords.pl &lt;new_gff_minus_internal_stop_peptides.gff&gt; &gt; &lt;new_gff_with_no_errors.gff&gt;
</pre>



<p>
and then remove the leading Ns from the sequences
</p></div>

</div>

<div id="outline-container-3.4" class="outline-3">
<h3 id="sec-3.4">convert files to NCBI-ready with my messy table scripts </h3>
<div class="outline-text-3" id="text-3.4">


<p>
the first script takes the gff as is, and puts the information into table formt
</p>


<pre class="example">~/bin/katie_gff2tbl.pl &lt;new_gff_with_no_errors.gff&gt; &gt; &lt;genome_annot_from_gff.tbl-sub&gt;
</pre>



<p>
the second script takes a file called <code>gene_name_conversions.txt</code> and substitutes the correct gene names into the tbl file from the previous command.
check <i>c-strain-annotation.sub.txt</i> for an example of the format of this file, which is tab-delimited and expects a header line.
the subsequent command just changes gff identifiers to NCBI identifiers
</p>


<pre class="example">~/bin/katie_gff2tble2.pl &lt;genome_annot_from_gff.tbl-sub&gt; &gt; &lt;genome_annot_from_1gff.tbl&gt;
perl -pe 's/gene_id/protein_id/' &lt;genome_annot_from_1gff.tbl&gt; &gt; &lt;genome_annot_from_2gff.tbl&gt;
</pre>



</div>

</div>

<div id="outline-container-3.5" class="outline-3">
<h3 id="sec-3.5">change the gene names to ones that are permissible by NCBI. </h3>
<div class="outline-text-3" id="text-3.5">

<p>file to annotate the specific identifier assigned by NCBI. <code>newIDs.new.ncbi.txt</code>
look at <i>newIDs.new.txt</i> for an example on how to do it. remember, the prefix that 
you entered in the WGS submission will go at the beginning of your new gene IDs. In the 
example file, I have the number followed by a dash. USE AN UNDERSCORE, NOT A DASH.
</p>


<pre class="example">~/bin/change_IDs.pl &lt;genome_annot_from_2gff.tbl&gt; &gt; &lt;final.tbl&gt;
</pre>




</div>

</div>

<div id="outline-container-3.6" class="outline-3">
<h3 id="sec-3.6">miscellaneous help </h3>
<div class="outline-text-3" id="text-3.6">

<p>these were to clean up things listed as problematic in the discrepancy report
this example shows deleting weird names that muck up the translator like "AF443189_2" and "homolog" and "af376000_1"
in the .tbl annotations
</p>



<pre class="example">perl -pe 's/\"//g;s/AF443189_2//g;s/, putative//g;s/, //g;s/homolog$/-like protein/g;s/CPUR_[0-9]+//g;s/af376000_1//g;s/pf11_0240-like//g;' c-strain.contigs.rename.tbl &gt; c-strain.contigs.tbl
</pre>




</div>
</div>

</div>

<div id="outline-container-4" class="outline-2">
<h2 id="sec-4">Step 4 is to submit </h2>
<div class="outline-text-2" id="text-4">


</div>

<div id="outline-container-4.1" class="outline-3">
<h3 id="sec-4.1">run most recent version of tbl2asn. </h3>
<div class="outline-text-3" id="text-4.1">

<p>Need to go download mac.tbl2asn or whatever it is called
Enec.sbt came from the NCBI website creating an sbt file for me at
<a href="http://www.ncbi.nlm.nih.gov/WebSub/template.cgi">http://www.ncbi.nlm.nih.gov/WebSub/template.cgi</a>
</p>
</div>

<div id="outline-container-4.1.1" class="outline-4">
<h4 id="sec-4.1.1">change things that are different, like paired-ends, organism name, isolate name, check your options, etc. </h4>
<div class="outline-text-4" id="text-4.1.1">




<pre class="example">./mac.tbl2asn -i &lt;&gt; -t Enec.sbt -M n -a r10k -j"[organism=Erysiphe necator][isolate=c]" -V v -l paired-ends
</pre>




</div>
</div>

</div>

<div id="outline-container-4.2" class="outline-3">
<h3 id="sec-4.2">check errorsummary.val for fatal errors </h3>
<div class="outline-text-3" id="text-4.2">

</div>

</div>

<div id="outline-container-4.3" class="outline-3">
<h3 id="sec-4.3">submit .sqn and .fas file </h3>
<div class="outline-text-3" id="text-4.3">

</div>
</div>
</div>
<div id="postamble">
<p class="author"> Author: Katherine Amrine
<a href="mailto:kc.amrine@gmail.com">&lt;kc.amrine@gmail.com&gt;</a>
</p>
<p class="date"> Date: 2014-09-08 15:59:53 PDT</p>
<p class="creator">HTML generated by org-mode 6.33x in emacs 23</p>
</div>
</div>
</body>
</html>
