<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
               "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
lang="en" xml:lang="en">
<head>
<title>Starter cookbook to processing genome w/ annotation pacakges for NCBI</title>
<meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1"/>
<meta name="generator" content="Org-mode"/>
<meta name="generated" content="2015-02-10 09:33:44 PST"/>
<meta name="author" content="Katherine Amrine"/>
<meta name="description" content=""/>
<meta name="keywords" content=""/>
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  html { font-family: Times, serif; font-size: 12pt; }
  .title  { text-align: center; }
  .todo   { color: red; }
  .done   { color: green; }
  .tag    { background-color: #add8e6; font-weight:normal }
  .target { }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  p.verse { margin-left: 3% }
  pre {
	border: 1pt solid #AEBDCC;
	background-color: #F3F5F7;
	padding: 5pt;
	font-family: courier, monospace;
        font-size: 90%;
        overflow:auto;
  }
  table { border-collapse: collapse; }
  td, th { vertical-align: top; }
  dt { font-weight: bold; }
  div.figure { padding: 0.5em; }
  div.figure p { text-align: center; }
  .linenr { font-size:smaller }
  .code-highlighted {background-color:#ffff00;}
  .org-info-js_info-navigation { border-style:none; }
  #org-info-js_console-label { font-size:10px; font-weight:bold;
                               white-space:nowrap; }
  .org-info-js_search-highlight {background-color:#ffff00; color:#000000;
                                 font-weight:bold; }
  /*]]>*/-->
</style>
<script type="text/javascript">
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">

<h1 class="title">Starter cookbook to processing genome w/ annotation pacakges for NCBI</h1>

<p>Alrighty, here is my step-by-step and explanation for what I did to
do a WGS annotated genome submission for NCBI
</p>

<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#sec-1">1 Notify NCBI you will be submitting data </a>
<ul>
<li><a href="#sec-1.1">1.1 create a bioproject submission per genome. </a></li>
<li><a href="#sec-1.2">1.2 Next create a biosample submission </a></li>
<li><a href="#sec-1.3">1.3 Next start the WGS submission process </a>
<ul>
<li><a href="#sec-1.3.1">1.3.1 WHAT YOU WILL NEED </a></li>
<li><a href="#sec-1.3.2">1.3.2 WHAT I GIVE YOU </a></li>
</ul></li>
</ul>
</li>
<li><a href="#sec-2">2 Pre-process the peptide files </a>
<ul>
<li><a href="#sec-2.1">2.1 isolate the peptide sequences with internal stops (check number of seqs) </a></li>
<li><a href="#sec-2.2">2.2 remove peptide sequences with internal stops (confirm you are missing seqs) </a></li>
<li><a href="#sec-2.3">2.3 remove annotated genes with internal stops from gff file. </a></li>
<li><a href="#sec-2.4">2.4 make sure there are no sequences in your gff file. It should only be coordinates for the following scripts. </a></li>
</ul>
</li>
<li><a href="#sec-3">3 Fix the scaffolds files </a>
<ul>
<li><a href="#sec-3.1">3.1 remove trailing Ns (no one cares where these are, just truncate them) </a></li>
<li><a href="#sec-3.2">3.2 find sequences with leading n's and their numbers, put into file </a></li>
<li><a href="#sec-3.3">3.3 change coordinates for leading N sequences in gff file </a></li>
<li><a href="#sec-3.4">3.4 and then remove the leading Ns from the sequences </a></li>
<li><a href="#sec-3.5">3.5 convert files to NCBI-ready with my messy table scripts </a></li>
<li><a href="#sec-3.6">3.6 change the gene names to ones that are permissible by NCBI. </a></li>
<li><a href="#sec-3.7">3.7 miscellaneous help </a></li>
</ul>
</li>
<li><a href="#sec-4">4 Prepare files to submit </a>
<ul>
<li><a href="#sec-4.1">4.1 make a .sbt file online </a></li>
<li><a href="#sec-4.2">4.2 get tbl2asn script from NCBI </a></li>
<li><a href="#sec-4.3">4.3 run most recent version of tbl2asn. change things that are different, like paired-ends, organism name, isolate name, check your options, -a has to do with gap sizes, etc. </a></li>
<li><a href="#sec-4.4">4.4 check errorsummary.val for fatal errors </a></li>
</ul>
</li>
<li><a href="#sec-5">5 submit .sqn and .fas file </a></li>
<li><a href="#sec-6">6 Fix the errors they find </a>
<ul>
<li><a href="#sec-6.1">6.1 adapter contamination </a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-1" class="outline-2">
<h2 id="sec-1"><span class="section-number-2">1</span> Notify NCBI you will be submitting data </h2>
<div class="outline-text-2" id="text-1">



</div>

<div id="outline-container-1.1" class="outline-3">
<h3 id="sec-1.1"><span class="section-number-3">1.1</span> create a bioproject submission per genome. </h3>
<div class="outline-text-3" id="text-1.1">

<p>Specify that you will be submitting WGS, Annotated, SRA, etc. I believe here, you will 
use an identifier to identify your genome that will act as the base of your locus tags. 
For Powdery mildew, it was automatically assigned as "EV44." Pick something meaningful
if you don't want one automatically assigned. When you get down to the 
<code>newIDs.new.ncbi.txt</code> file I ask for in Step3, this prefix is what needs to go in this file.
</p>
</div>

</div>

<div id="outline-container-1.2" class="outline-3">
<h3 id="sec-1.2"><span class="section-number-3">1.2</span> Next create a biosample submission </h3>
<div class="outline-text-3" id="text-1.2">

<p>include details on your sample
</p>
</div>

</div>

<div id="outline-container-1.3" class="outline-3">
<h3 id="sec-1.3"><span class="section-number-3">1.3</span> Next start the WGS submission process </h3>
<div class="outline-text-3" id="text-1.3">

<p>also, they will ask you for the type of evidence that you have for the strings of 
N's within sequences. First off, N's do NOT occur as singletons, so they are not 
ambiguous bases with CLC. Ns refer to gaps of the size of the string of Ns if they ask you, but
if you do have Ns as ambiguous bases, you need to specify this. 
</p>
</div>

<div id="outline-container-1.3.1" class="outline-4">
<h4 id="sec-1.3.1"><span class="section-number-4">1.3.1</span> WHAT YOU WILL NEED </h4>
<div class="outline-text-4" id="text-1.3.1">

<p>-fasta file with scaffolds with concise names
this means that you really need to slim down names to as simple of
identifiers as possible. 
</p>
<p>
-gff version 2 file (with the scripts I provide here) devoid of sequences
you don't need version 2, but I used it and wrote my scripts to
use this type of file. The GFF file gene names should match your 
gene names in your fasta file.
</p>
<p>
-install of BIOPERL and <a href="http://search.cpan.org/~dhard/FAST-1.0/">FASTools</a>
</p>
</div>

</div>

<div id="outline-container-1.3.2" class="outline-4">
<h4 id="sec-1.3.2"><span class="section-number-4">1.3.2</span> WHAT I GIVE YOU </h4>
<div class="outline-text-4" id="text-1.3.2">

<ul>
<li id="sec-1.3.2.1">a crappy script to do the first conversion of the gff file to the tbl <br/>
</li>
<li id="sec-1.3.2.2">file required for the tbl2asn script that will ultimately make your  asn submission file <br/>
</li>
<li id="sec-1.3.2.3">another crappy script to substitute in the names necessary <br/>

</li>
</ul>
</div>
</div>
</div>

</div>

<div id="outline-container-2" class="outline-2">
<h2 id="sec-2"><span class="section-number-2">2</span> Pre-process the peptide files </h2>
<div class="outline-text-2" id="text-2">


</div>

<div id="outline-container-2.1" class="outline-3">
<h3 id="sec-2.1"><span class="section-number-3">2.1</span> isolate the peptide sequences with internal stops (check number of seqs) </h3>
<div class="outline-text-3" id="text-2.1">

<p>needs: peptide fasta file
</p>
<p>
returns: fasta file with internal stop sequences
</p>


<pre class="example">fasgrep -s "[^A-Z].+" &lt;peptide_file.fas&gt; &gt; &lt;file_with_internal_stop_sequences.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-2.2" class="outline-3">
<h3 id="sec-2.2"><span class="section-number-3">2.2</span> remove peptide sequences with internal stops (confirm you are missing seqs) </h3>
<div class="outline-text-3" id="text-2.2">

<p>needs: peptide fasta file
</p>
<p>
returns: fasta file without the internal stop sequences
</p>


<pre class="example">fasgrep -vs "[^A-Z].+" &lt;peptide_file.fas&gt; &gt; &lt;peptiled_file_minus_internal_stops.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-2.3" class="outline-3">
<h3 id="sec-2.3"><span class="section-number-3">2.3</span> remove annotated genes with internal stops from gff file. </h3>
<div class="outline-text-3" id="text-2.3">

<p>needs: peptide fasta file without the internal stop sequences; gff file
</p>
<p>
returns: gff file without genes with internal stops
</p>
<p>
the second command (cut -c 2-) will get everything but the "&gt;" character, and the 
third command (cut -f1,2 -d".") is specific to the e-necator processing. Our gene
names consisted of the species, then gene name separated by periods. There was a third
field with ".t1" which specifies only the transcript in our gff files. We remove this
because it would only pick up the mRNA line in the gff file. You need to see what is 
unique about your gene names and modify this. The fourth command (perl code) adds a 
regex end-of-word wildcard. This only works if your gene names in your fasta peptide file
match the gene names in your gff file. You should check this. 
(confirm correct unique ids are removed after this step)
</p>



<pre class="example">egrep -f &lt;(egrep "&gt;" &lt;peptide_file_minus_internal_stops.fas&gt; | cut -c 2- |cut -f1,2 -d"." | perl -pe "s/\n/\\\&gt;\n/") &lt;original_gff_file.gff&gt; &gt; &lt;new_gff_minus_inernal_stop_peptides.gff&gt;
</pre>



</div>

</div>

<div id="outline-container-2.4" class="outline-3">
<h3 id="sec-2.4"><span class="section-number-3">2.4</span> make sure there are no sequences in your gff file. It should only be coordinates for the following scripts. </h3>
<div class="outline-text-3" id="text-2.4">

</div>
</div>

</div>

<div id="outline-container-3" class="outline-2">
<h2 id="sec-3"><span class="section-number-2">3</span> Fix the scaffolds files </h2>
<div class="outline-text-2" id="text-3">


</div>

<div id="outline-container-3.1" class="outline-3">
<h3 id="sec-3.1"><span class="section-number-3">3.1</span> remove trailing Ns (no one cares where these are, just truncate them) </h3>
<div class="outline-text-3" id="text-3.1">

<p>needs: original scaffold file
</p>
<p>
returns: scaffold file without trailing ends
</p>


<pre class="example">fassub -s 'N{1,}$' '' &lt;scaffolds_file_original.fas&gt; &gt; &lt;scaffolds_no_trailing.fas&gt;
</pre>



</div>

</div>

<div id="outline-container-3.2" class="outline-3">
<h3 id="sec-3.2"><span class="section-number-3">3.2</span> find sequences with leading n's and their numbers, put into file </h3>
<div class="outline-text-3" id="text-3.2">

<p>needs: scaffold file from previous step
</p>
<p>
returns: text file with list of sequences that have trailing N's, and their number
</p>


<pre class="example">fasgrep -s "^N" &lt;scaffolds_no_trailing.fas&gt; | perl -ne 'BEGIN{$start=0;}if(/^\&gt;/){$finish=0;chomp;$seqname=$_;$start=1;$length=0;}elsif($start==1 &amp;&amp; $finish==0){/(^N{1,})/;$length += length($1);if(/[ATCG]/){ print join"",$seqname,"\t",$length,"\n";$start=0;$finish=1;}}' | sort | uniq &gt; leading_Ns.txt
</pre>



</div>

</div>

<div id="outline-container-3.3" class="outline-3">
<h3 id="sec-3.3"><span class="section-number-3">3.3</span> change coordinates for leading N sequences in gff file </h3>
<div class="outline-text-3" id="text-3.3">

<p>needs: file specifying leading N's
</p>
<p>
returns: gff file adjusting coordinats for genes on scaffolds where leading Ns were removed
</p>
<p>
for this script <i>changecoords.pl</i>, you need the file <code>leading_Ns.txt</code>
but it takes the gff file and fixes all the coordinates
</p>


<pre class="example">./change_coords.pl &lt;new_gff_minus_internal_stop_peptides.gff&gt; &gt; &lt;new_gff_with_no_errors.gff&gt;
</pre>



</div>

</div>

<div id="outline-container-3.4" class="outline-3">
<h3 id="sec-3.4"><span class="section-number-3">3.4</span> and then remove the leading Ns from the sequences </h3>
<div class="outline-text-3" id="text-3.4">

<p>needs: scaffold file edited from aboce
</p>
<p>
returns: scaffold file now with removed leading Ns 
</p>



<pre class="example">fassub -s '^N{1,}' '' &lt;scaffolds_no_trailing.fas&gt; &lt;scaffolds_no_trailing_leading.fas&gt; 
</pre>




</div>

</div>

<div id="outline-container-3.5" class="outline-3">
<h3 id="sec-3.5"><span class="section-number-3">3.5</span> convert files to NCBI-ready with my messy table scripts </h3>
<div class="outline-text-3" id="text-3.5">

<p>needs: processed gff file from previous steps, and a conversion file with the old gene names, and the new gene names that are NCBI-specific
</p>
<p>
returns: NCBI tbl file with NCBI-specific identifiers for proteins
</p>

<p>
the first script takes the gff as is, and puts the information into table formt
</p>


<pre class="example">~/bin/katie_gff2tbl.pl &lt;new_gff_with_no_errors.gff&gt; &gt; &lt;genome_annot_from_gff.tbl-sub&gt;
</pre>



<p>
the second script takes a file called <code>gene_name_conversions.txt</code> and substitutes the correct annotations into the tbl file from the previous command.
check <i>c-strain-annotation.sub.txt</i> for an example of the format of this file, which is tab-delimited and expects a header line.
the subsequent command just changes gff identifiers to NCBI identifiers
</p>


<pre class="example">/bin/katie_gff2tble2.pl &lt;genome_annot_from_gff.tbl-sub&gt; &gt; &lt;genome_annot_from_1gff.tbl&gt;
perl -pe 's/gene_id/protein_id/' &lt;genome_annot_from_1gff.tbl&gt; &gt; &lt;genome_annot_from_2gff.tbl&gt;
</pre>



</div>

</div>

<div id="outline-container-3.6" class="outline-3">
<h3 id="sec-3.6"><span class="section-number-3">3.6</span> change the gene names to ones that are permissible by NCBI. </h3>
<div class="outline-text-3" id="text-3.6">

<p>needs: .tbl file from previous step; file to annotate the specific identifier assigned by NCBI. <code>newIDs.new.ncbi.txt</code>
look at <i>newIDs.new.txt</i> for an example on how to do it. remember, the prefix that 
you entered in the WGS submission will go at the beginning of your new gene IDs. In the 
example file, I have the number followed by a dash. USE AN UNDERSCORE, NOT A DASH.
</p>
<p>
returns: ncbi-formatted tbl file
</p>


<pre class="example">~/bin/change_IDs.pl &lt;genome_annot_from_2gff.tbl&gt; &gt; &lt;final.tbl&gt;
</pre>




</div>

</div>

<div id="outline-container-3.7" class="outline-3">
<h3 id="sec-3.7"><span class="section-number-3">3.7</span> miscellaneous help </h3>
<div class="outline-text-3" id="text-3.7">

<p>these were to clean up things listed as problematic in the discrepancy report
this example shows deleting weird names that muck up the translator like "AF443189_2" and "homolog" and "af376000_1"
in the .tbl annotations
</p>



<pre class="example">perl -pe 's/\"//g;s/AF443189_2//g;s/, putative//g;s/, //g;s/homolog$/-like protein/g;s/CPUR_[0-9]+//g;s/af376000_1//g;s/pf11_0240-like//g;' c-strain.contigs.rename.tbl &gt; c-strain.contigs.tbl
</pre>




</div>
</div>

</div>

<div id="outline-container-4" class="outline-2">
<h2 id="sec-4"><span class="section-number-2">4</span> Prepare files to submit </h2>
<div class="outline-text-2" id="text-4">


</div>

<div id="outline-container-4.1" class="outline-3">
<h3 id="sec-4.1"><span class="section-number-3">4.1</span> make a .sbt file online </h3>
<div class="outline-text-3" id="text-4.1">

<p>Enec.sbt came from the NCBI website creating an sbt file for me at
<a href="http://www.ncbi.nlm.nih.gov/WebSub/template.cgi">http://www.ncbi.nlm.nih.gov/WebSub/template.cgi</a>
</p>
</div>

</div>

<div id="outline-container-4.2" class="outline-3">
<h3 id="sec-4.2"><span class="section-number-3">4.2</span> get tbl2asn script from NCBI </h3>
<div class="outline-text-3" id="text-4.2">

<p>Need to go download mac.tbl2asn or whatever it is called
</p>
</div>

</div>

<div id="outline-container-4.3" class="outline-3">
<h3 id="sec-4.3"><span class="section-number-3">4.3</span> run most recent version of tbl2asn. change things that are different, like paired-ends, organism name, isolate name, check your options, -a has to do with gap sizes, etc. </h3>
<div class="outline-text-3" id="text-4.3">

<p>this script looks for a .tbl file in your current directory
</p>


<pre class="example">./mac.tbl2asn -i &lt;&gt; -t Enec.sbt -M n -a r10k -j"[organism=Erysiphe necator][isolate=c]" -V v -l paired-ends
</pre>




</div>

</div>

<div id="outline-container-4.4" class="outline-3">
<h3 id="sec-4.4"><span class="section-number-3">4.4</span> check errorsummary.val for fatal errors </h3>
<div class="outline-text-3" id="text-4.4">

</div>
</div>

</div>

<div id="outline-container-5" class="outline-2">
<h2 id="sec-5"><span class="section-number-2">5</span> <a href="#color:red">submit .sqn and .fas file</a> </h2>
<div class="outline-text-2" id="text-5">

</div>

</div>

<div id="outline-container-6" class="outline-2">
<h2 id="sec-6"><span class="section-number-2">6</span> Fix the errors they find </h2>
<div class="outline-text-2" id="text-6">


</div>

<div id="outline-container-6.1" class="outline-3">
<h3 id="sec-6.1"><span class="section-number-3">6.1</span> adapter contamination </h3>
<div class="outline-text-3" id="text-6.1">

<p>You will likely have some adapter contamination they find with your submission. To fix this, Copy the lines from the email
NCBI sends out into a file named <i>adapters_to_remove.txt</i> (check this file for an example of the input). The next script is still in very raw form. You want to edit the lines
that specify your input files and your output files within the script. The standard output of this script is the scaffolds in 
fasta format. A new gff file will be created within the script. Bioperl is required. 
</p>



<pre class="example">./remove_NCBI_adaptors.pl &gt; &lt;new_trimmed_scaffolds.fas&gt;
</pre>



</div>
</div>
</div>
<div id="postamble">
<p class="author"> Author: Katherine Amrine
<a href="mailto:kc.amrine@gmail.com">&lt;kc.amrine@gmail.com&gt;</a>
</p>
<p class="date"> Date: 2015-02-10 09:33:44 PST</p>
<p class="creator">HTML generated by org-mode 6.33x in emacs 23</p>
</div>
</div>
</body>
</html>
